### DISCUSSION TITLE
Documentation: Multi-boot & ISO boot

### DISCUSSION THREAD

**Discussion by Beeblebrox-BSD** [zbm-dev/zfsbootmenu#567] [Feature Requests]
I was under the impression that zfsbootmenu could boot multiple OS with placement under zroot/ROOT/{ID}

I assume that zfsbootmenu would also be able to boot live ISO images - something Grub usually does well.

Can`t find documentation for either use case.

**Comment by classabbyamp** [comment]
> I was under the impression that zfsbootmenu could boot multiple OS with placement under zroot/ROOT/{ID}

they don't need to be placed at that exact dataset hierarchy but yes. it's not much different than doing an initial install except you don't need to create the pool. The [Bootenvs and You guide](https://docs.zfsbootmenu.org/en/v2.3.x/guides/general/bootenvs-and-you.html) talks about how ZBM discovers OSes. 

> I assume that zfsbootmenu would also be able to boot live ISO images - something Grub usually does well.

It cannot (that's out of ZBM's scope), but you can use rEFInd as an initial bootloader that launches ZBM and I think that supports launching ISOs.

**Comment by zdykstra** [comment]
ZFSBootMenu doesn't support booting an ISO - there's existing tooling (e.g. Ventoy) that does that already.

There's no real documentation needed for multiple boot environments beyond https://docs.zfsbootmenu.org/en/v2.3.x/guides/general/bootenvs-and-you.html . Create as many as you'd like, then boot them!

**Comment by Beeblebrox-BSD** [comment]
Hi, thank you both for your comments.

* The ISO issue is understood.
* I went through the documents before posting and after the suggestion that the solution for multi-boot is already there, but I still fail to understand several things:
1. What happens to second / third install on separate dataset with `zpool set bootfs=zroot/ROOT/${ID} zroot` ? Should this step be omitted and just keep the initial flag intact?
2. With encrypted zpool, each distro initramfs is rebuilt with the key embedded as I understand. What about
`zfs set org.zfsbootmenu:keysource="zroot/ROOT/${ID}" zroot`
3. Since the EFI setup was already done, I assume the step for "Configure EFI boot entries" should be ignored?
4.  Finally, how is ZBM made aware of the additional boot option? Does ZBM detect automatically from ZFS dataset flags `org.zfsbootmenu:active`?

Thank you.

**Reply by ahesford** [reply]
1. You set the `bootfs` pool property to point to whichever boot environment you want to be selected by default. Change it whenever you like.
2. If you use ZBM key caching, point `org.zfsbootmenu:keysource` **on the encryption root** to any of the filesystems that contains the keys. If each environment is encrypted separately, presumably each environment is its own encryption root and should have its own key source pointing to itself. If you copy the keys to every boot environment and have a single encryption root, point the common encryption root to any of the environments that has a copy. If you mount a separate filesystem as a standalone source for keys, point your key source there.
3. Correct; there is no need to reconfigure your firmware if you are not changing the ZFSBootMenu file name.
4. ZFSBootMenu dynamically scans your pools according to the process described in "Boot Environments and You". The `active` property is only useful to opt **out** of detecting a boot environment with `mountpoint=/`, or to opt **in** to detecting a boot environment with `mountpoint=legacy`. In general, you should not need to set that property anywhere.

**Comment by Beeblebrox-BSD** [comment]
Thanks for the tips. I got dual boot working per your guidance.
